docs/brotli.1
.\" Automatically generated by Pandoc 2.7.3
.\" Manually corrected
.\"
.TH "brotli" "1" "August 29 2023" "brotli 1.1.0" "User Manual"
.hy
.SH NAME
.PP
brotli(1) -- brotli, brcat, unbrotli - compress or decompress files
.SH SYNOPSIS
.PP
\f[B]brotli\f[R] [\f[I]OPTION|FILE\f[R]]\&...
.PP
\f[B]brcat\f[R] is equivalent to \f[B]brotli --decompress --concatenated
--stdout\f[R]
.PP
\f[B]unbrotli\f[R] is equivalent to \f[B]brotli --decompress\f[R]
.SH DESCRIPTION
.PP
\f[B]brotli\f[R] is a generic-purpose lossless compression algorithm
that compresses data using a combination of a modern variant of the
\f[B]LZ77\f[R] algorithm, Huffman coding and 2-nd order context
modeling, with a compression ratio comparable to the best currently
available general-purpose compression methods.
It is similar in speed with deflate but offers more dense compression.
.PP
\f[B]brotli\f[R] command line syntax similar to \f[B]gzip (1)\f[R] and
\f[B]zstd (1)\f[R].
Unlike \f[B]gzip (1)\f[R], source files are preserved by default.
It is possible to remove them after processing by using the
\f[B]--rm\f[R] \f[I]option\f[R].
.PP
Arguments that look like \[lq]\f[B]--name\f[R]\[rq] or
\[lq]\f[B]--name=value\f[R]\[rq] are \f[I]options\f[R].
Every \f[I]option\f[R] has a short form \[lq]\f[B]-x\f[R]\[rq] or
\[lq]\f[B]-x value\f[R]\[rq].
Multiple short form \f[I]options\f[R] could be coalesced:
.IP \[bu] 2
\[lq]\f[B]--decompress --stdout --suffix=.b\f[R]\[rq] works the same as
.IP \[bu] 2
\[lq]\f[B]-d -s -S .b\f[R]\[rq] and
.IP \[bu] 2
\[lq]\f[B]-dsS .b\f[R]\[rq]
.PP
\f[B]brotli\f[R] has 3 operation modes:
.IP \[bu] 2
default mode is compression;
.IP \[bu] 2
\f[B]--decompress\f[R] option activates decompression mode;
.IP \[bu] 2
\f[B]--test\f[R] option switches to integrity test mode; this option is
equivalent to \[lq]\f[B]--decompress --stdout\f[R]\[rq] except that the
decompressed data is discarded instead of being written to standard
output.
.PP
Every non-option argument is a \f[I]file\f[R] entry.
If no \f[I]files\f[R] are given or \f[I]file\f[R] is
\[lq]\f[B]-\f[R]\[rq], \f[B]brotli\f[R] reads from standard input.
All arguments after \[lq]\f[B]--\f[R]\[rq] are \f[I]file\f[R] entries.
.PP
Unless \f[B]--stdout\f[R] or \f[B]--output\f[R] is specified,
\f[I]files\f[R] are written to a new file whose name is derived from the
source \f[I]file\f[R] name:
.IP \[bu] 2
when compressing, a suffix is appended to the source filename to get the
target filename
.IP \[bu] 2
when decompressing, a suffix is removed from the source filename to get
the target filename
.PP
Default suffix is \f[B].br\f[R], but it could be specified with
\f[B]--suffix\f[R] option.
.PP
Conflicting or duplicate \f[I]options\f[R] are not allowed.
.SH OPTIONS
.IP \[bu] 2
\f[B]-#\f[R]: compression level (0-9); bigger values cause denser, but
slower compression
.IP \[bu] 2
\f[B]-c\f[R], \f[B]--stdout\f[R]: write on standard output
.IP \[bu] 2
\f[B]-d\f[R], \f[B]--decompress\f[R]: decompress mode
.IP \[bu] 2
\f[B]-f\f[R], \f[B]--force\f[R]: force output file overwrite
.IP \[bu] 2
\f[B]-h\f[R], \f[B]--help\f[R]: display this help and exit
.IP \[bu] 2
\f[B]-j\f[R], \f[B]--rm\f[R]: remove source file(s); \f[B]gzip
(1)\f[R]-like behaviour
.IP \[bu] 2
\f[B]-k\f[R], \f[B]--keep\f[R]: keep source file(s); \f[B]zstd
(1)\f[R]-like behaviour
.IP \[bu] 2
\f[B]-s\f[R], \f[B]--squash\f[R]: remove destination file if larger than source
.IP \[bu] 2
\f[B]-n\f[R], \f[B]--no-copy-stat\f[R]: do not copy source file(s)
attributes
.IP \[bu] 2
\f[B]-o FILE\f[R], \f[B]--output=FILE\f[R] output file; valid only if
there is a single input entry
.IP \[bu] 2
\f[B]-q NUM\f[R], \f[B]--quality=NUM\f[R]: compression level (0-11);
bigger values cause denser, but slower compression
.IP \[bu] 2
\f[B]-t\f[R], \f[B]--test\f[R]: test file integrity mode
.IP \[bu] 2
\f[B]-v\f[R], \f[B]--verbose\f[R]: increase output verbosity
.IP \[bu] 2
\f[B]-w NUM\f[R], \f[B]--lgwin=NUM\f[R]: set LZ77 window size (0, 10-24)
(default: 24); window size is \f[B](pow(2, NUM) - 16)\f[R]; 0 lets
compressor decide over the optimal value; bigger windows size improve
density; decoder might require up to window size memory to operate
.IP \[bu] 2
\f[B]--large_window=NUM\f[R]: use incompatible large-window  brotli
bitstream with window size (0, 10-30) WARNING: this format is not
compatible with brotli RFC 7932 and may not be decodable with regular
brotli decoders
.IP \[bu] 2
\f[B]-C B64\f[R], \f[B]--comment=B64\f[R]: set comment; argument is
base64-decoded first; when decoding: check stream comment; when
encoding: embed comment (fingerprint)
.IP \[bu] 2
\f[B]-D FILE\f[R], \f[B]--dictionary=FILE\f[R]: use FILE as raw (LZ77)
dictionary; same dictionary MUST be used both for compression and
decompression
.IP \[bu] 2
\f[B]-K\f[R], \f[B]--concatenated\f[R]: when decoding, allow
concatenated brotli streams as input
.IP \[bu] 2
\f[B]-S SUF\f[R], \f[B]--suffix=SUF\f[R]: output file suffix (default:
\f[B].br\f[R])
.IP \[bu] 2
\f[B]-V\f[R], \f[B]--version\f[R]: display version and exit
.IP \[bu] 2
\f[B]-Z\f[R], \f[B]--best\f[R]: use best compression level (default);
same as \[lq]\f[B]-q 11\f[R]\[rq]
.SH SEE ALSO
.PP
\f[B]brotli\f[R] file format is defined in RFC
7932 (https://www.ietf.org/rfc/rfc7932.txt).
.PP
\f[B]brotli\f[R] is open-sourced under the MIT
License (https://opensource.org/licenses/MIT).
.PP
Mailing list: https://groups.google.com/forum/#!forum/brotli
.SH BUGS
.PP
Report bugs at: https://github.com/google/brotli/issues


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

### Added
 - decoder: `BrotliDecoderAttachDictionary`
 - decoder: `BrotliDecoderOnFinish` callback behind `BROTLI_REPORTING`
 - decoder: `BrotliDecoderSetMetadataCallbacks`
 - encoder: `BrotliEncoderPrepareDictionary`,
            `BrotliEncoderDestroyPreparedDictionary`,
            `BrotliEncoderAttachPreparedDictionary`
 - decoder: `BrotliEncoderOnFinish` callback behind `BROTLI_REPORTING`
 - common: `BrotliSharedDictionaryCreateInstance`,
           `BrotliSharedDictionaryDestroyInstance`,
           `BrotliSharedDictionaryAttach`
 - CLI: `--dictionary` option
 - java: encoder wrapper: `Parameters.mode`
 - java: `Brotli{Input|Output}Stream.attachDictionary`
 - java: wrapper: partial byte array input
 - typescript: decoder (transpiled from Java)

### Removed
 - build: `BROTLI_BUILD_PORTABLE` option

### Fixed
 - java: JNI decoder failed sometimes on power of 2 payloads

### Improved
 - java / js: smaller decoder footprint
 - decoder: faster decoding
 - encoder: faster encoding
 - encoder: smaller stack frames


## [1.0.9] - 2020-08-27

Re-release of 1.0.8.


## [1.0.8] - 2020-08-27

### SECURITY
 - CVE-2020-8927: potential overflow when input chunk is >2GiB

### Added
 - encoder: `BROTLI_PARAM_STREAM_OFFSET`

### Improved
 - CLI: better reporting
 - CLI: workaround for "lying feof"
 - java: faster decoding
 - java: support "large window"
 - encoder: use less memory
 - release: filter sources for the tarball


## [1.0.7] - 2018-10-23

### Improved
 - decoder: faster decoding on ARM CPU


## [1.0.6] - 2018-09-13

### Fixed
 - build: AutoMake and CMake build
 - java: JDK 8<->9 incompatibility


## [1.0.5] - 2018-06-27

### Added
 - scripts: extraction of static dictionary from RFC

### Improved
 - encoder: better compression at quality 1
 - encoder: better compression with "large window"


## [1.0.4] - 2018-03-29

### Added
 - encoder: `BROTLI_PARAM_NPOSTFIX`, `BROTLI_PARAM_NDIRECT`
 - CLI: `--large_window` option

### Improved
 - encoder: better compression


## [1.0.3] - 2018-03-02

### Added
 - decoder: `BROTLI_DECODER_PARAM_LARGE_WINDOW` enum
 - encoder: `BROTLI_PARAM_LARGE_WINDOW` enum
 - java: `BrotliInputStream.setEager`

### Fixed
 - build: AutoMake build in some environments
 - encoder: fix one-shot q=10 1-byte input compression

### Improved
 - encoder: better font compression


## [1.0.2] - 2017-11-28

### Added
 - build: AutoMake
 - research: better dictionary generators


## [1.0.1] - 2017-09-22

### Changed
 - clarifications in `README.md`


## [1.0.0] - 2017-09-20

### Added
 - decoder: `BrotliDecoderSetParameter`
 - csharp: decoder (transpiled from Java)
 - java: JNI wrappers
 - javascript: decoder (transpiled from Java)
 - python: streaming decompression
 - research: dictionary generator

### Changed
 - CLI: rename `bro` to `brotli`

### Removed
 - decoder: `BrotliDecoderSetCustomDictionary`
 - encoder: `BrotliEncoderSetCustomDictionary`

### Improved
 - java: faster decoding
 - encoder: faster compression


## [0.6.0] - 2017-04-10

### Added
 - CLI: `--no-copy-stat option
 - java: pure java decoder
 - build: fuzzers
 - research: `brotlidump` tool to explore brotli streams
 - go: wrapper

### Removed
 - decoder: API with plain `Brotli` prefix

### Deprecated
 - encoder: `BrotliEncoderInputBlockSize`, `BrotliEncoderCopyInputToRingBuffer`,
            `BrotliEncoderWriteData`

### Improved
 - encoder: faster compression
 - encoder: denser compression
 - decoder: faster decompression
 - python: release GIL
 - python: use zero-copy API


## [0.5.2] - 2016-08-11

### Added
 - common: `BROTLI_BOOL`, `BROTLI_TRUE`, `BROTLI_FALSE`
 - decoder: API with `BrotliDecoder` prefix instead of plain `Brotli`
 - build: Bazel, CMake

### Deprecated
 - decoder: API with plain `Brotli` prefix

### Changed
 - boolean argument / result types are re-branded as `BROTLI_BOOL`

### Improved
 - build: reduced amount of warnings in various build environments
 - encoder: faster compression
 - encoder: lower memory usage


## [0.5.0] - 2016-06-15

### Added
 - common: library has been assembled from shared parts of decoder and encoder
 - encoder: C API

### Removed
 - encoder: C++ API


## [0.4.0] - 2016-06-14

### Added
 - encoder: faster compression modes (quality 0 and 1)
 - decoder: `BrotliGetErrorCode`, `BrotliErrorString` and
            `BROTLI_ERROR_CODES_LIST`

### Removed
 - decoder: deprecated streaming API (using `BrotliInput`)

### Fixed
 - decoder: possible pointer underflow

### Improved
 - encoder: faster compression


## [0.3.0] - 2015-12-22

### LICENSE
License have been upgraded to more permissive MIT.

### Added
 - CLI: `--window` option
 - `tools/version.h` file
 - decoder: low level streaming API
 - decoder: custom memory manager API

### Deprecated
 - decoder: streaming API using `BrotliInput` struct

### Fixed
 - decoder: processing of uncompressed blocks
 - encoder: possible division by zero

### Improved
 - encoder: faster decompression
 - build: more portable builds for various CPU architectures


## [0.2.0] - 2015-09-01

### Added
 - CLI: `--verbose` and `--repeat` options

### Fixed
 - decoder: processing of uncompressed blocks
 - encoder: block stitching on quality 10 / 11

### Improved
 - build: CI/CD integration
 - build: better test coverage
 - encoder: better compression of UTF-8 content
 - encoder: faster decompression


## [0.1.0] - 2015-08-11

Initial release.


python/brotli.py
# Copyright 2016 The Brotli Authors. All rights reserved.
#
# Distributed under MIT license.
# See file LICENSE for detail or copy at https://opensource.org/licenses/MIT

"""Functions to compress and decompress data using the Brotli library."""

import _brotli

# The library version.
__version__ = _brotli.__version__

# The compression mode.
MODE_GENERIC = _brotli.MODE_GENERIC
MODE_TEXT = _brotli.MODE_TEXT
MODE_FONT = _brotli.MODE_FONT

# The Compressor object.
Compressor = _brotli.Compressor

# The Decompressor object.
Decompressor = _brotli.Decompressor

# Compress a byte string.
def compress(string, mode=MODE_GENERIC, quality=11, lgwin=22, lgblock=0):
    """Compress a byte string.

    Args:
      string (bytes): The input data.
      mode (int, optional): The compression mode can be MODE_GENERIC (default),
        MODE_TEXT (for UTF-8 format text input) or MODE_FONT (for WOFF 2.0).
      quality (int, optional): Controls the compression-speed vs compression-
        density tradeoff. The higher the quality, the slower the compression.
        Range is 0 to 11. Defaults to 11.
      lgwin (int, optional): Base 2 logarithm of the sliding window size. Range
        is 10 to 24. Defaults to 22.
      lgblock (int, optional): Base 2 logarithm of the maximum input block size.
        Range is 16 to 24. If set to 0, the value will be set based on the
        quality. Defaults to 0.

    Returns:
      The compressed byte string.

    Raises:
      brotli.error: If arguments are invalid, or compressor fails.
    """
    compressor = Compressor(mode=mode, quality=quality, lgwin=lgwin,
                            lgblock=lgblock)
    return compressor.process(string) + compressor.finish()

# Decompress a compressed byte string.
decompress = _brotli.decompress

# Raised if compression or decompression fails.
error = _brotli.error
